### ISSUE TITLE
IfcMappedItem optimizations when item's StyledByItem count != 0

### ISSUE THREAD

**Issue by Stinkfist0** [IfcOpenShell/IfcOpenShell#143]
I'd have a structural model where the IfcMappedItems are not instantiated but unique meshes are generated for each case and the generated DAE or OBJ will be huge. AFAIK, there are currently known limitations in the IfcMappedItem optimization. I could share the the test file with you Thomas, if you think you have time to take a look. Or alternatively pointers for resolving this on my own are welcome too.

**Comment by jf---** [comment]
@Stinkfist0 , wavefront .obj doesnt support instances, right?
any idea how opencollada handles instances?
a more precise idea about the wished for outcome is probably useful...
interesting problem though...

**Comment by Stinkfist0** [comment]
Yeah, only ColladaSerializer utilizes this optimization currently. It has been in for some time already, see #47

**Comment by aothms** [comment]
@Stinkfist0  sorry for the slow reply. The thing to find out would be which of this long list of conditions [1] triggers the mapped item not to be processed. From the issue title it seems to be related to the item styles.

If the item really introduces a style different from its target representation, it cannot be re-used, as currently the processed items contain the associated styles. In DAE this is not strictly necessary as materials can be associated on an `object` level rather than a `mesh` level. But fixing this would be as far reaching I don't think is feasible on the current code base.

Alternatively if the item style is redundant, i.e. it assigns a style that is also assigned to the individual items of the target representation, than perhaps this can be detected.

Or, perhaps there is a number of styles assigned that is smaller than the total number of mapped instances. Meaning that within the set of instances, there are subsets that reuse each others processed representations. In which case perhaps you can group these subsets. For this though, probably the iteration mechanism has to be adapted.

It is not a completely trivial problem. Perhaps it would help if we can look at some diagram of the instances to see what could work in this particular situation.

[1] https://github.com/3drepo/IfcOpenShell/blob/e7e0fef1c394a32b6e3caf44933f8efdcb9958af/src/ifcgeom/IfcGeomIterator.h#L443

**Comment by Stinkfist0** [comment]
Just an update that the model's problem most likely is not with styled items and that was a false uneducated guess.

**Comment by aothms** [comment]
> Just an update that the model's problem most likely is not with styled items and that was a false uneducated guess ;)

Well, no, actually your intuition was correct, there are also `IfcStyledItems` associated to the `IfcMappedItems`, causing them to not being processed as instances, due to the difference in materials. In addition to that, the `MappingTarget` (=`IfcCartesianTransformationOperator3D`) also had a non identity transformation, which ends up getting multipled into the mesh vertex position, hence also not considered instances.

**Comment by Stinkfist0** [comment]
Ah right, so it was a combination of those two.
